### PR TITLE
Fix IPv6 TCP MSS issue

### LINKED ISSUE #56657: Incorrect IPv6 MSS size calculation leads to send failures
**Describe the bug**

In my setup, I'm using TCP connections with IPv6 through Openthread. MTU is by default 1280

What happens is that this line returns a correct MSS of 1220
https://github.com/zephyrproject-rtos/zephyr/blob/38d9da726644c8e82b3d8390ef85a175d63804ec/subsys/net/ip/tcp.c#L3368

But this line, sets it (incorrectly) to 1280.
https://github.com/zephyrproject-rtos/zephyr/blob/38d9da726644c8e82b3d8390ef85a175d63804ec/subsys/net/ip/tcp.c#L3372

This later causes this line to fail with OT_ERROR_PARSE (6), caused by Header parsing validation failing on too large message.
https://github.com/zephyrproject-rtos/zephyr/blob/ab410807bcc73fe9e88fb6c2dba393e2750aa4c1/modules/openthread/platform/radio.c#L445
It turns out that the message is 1340 bytes and does not fit into the MTU.

**To Reproduce**
Send an IPv6 TCP packet with size between 1221 and 1280.

**Expected behavior**
I expect the maximum MSS size for MTU 1280 to be 1220

**Impact**
I think this is a serious bug which probably impact more people than me. I can work around this by either patching my version (fix is trivial), or sending data with TCP_NODELAY and limited packet size.

**Environment (please complete the following information):**
Using main branch as of 2022-04-07

### PR DESCRIPTION
**Pull request by wirehell**
- Add an MTU check to loopback
- Fixing https://github.com/zephyrproject-rtos/zephyr/issues/56657
- Align IPv4 and IPv6 MSS calculation

### CODE REVIEW

File: tests/net/socket/tcp/src/main.c [hunk 1]
```diff
@@ -335,18 +335,37 @@ void tcp_server_block_thread(void *vps_sock, void *unused2, void *unused3)
 	test_close(new_sock);
 }
 
-void test_v4_send_recv_large_common(int tcp_nodelay)
+void test_send_recv_large_common(int tcp_nodelay, int family)
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:338]: Please split the test extension into a separate commit.

File: tests/net/socket/tcp/src/main.c [hunk 2]
```diff
@@ -416,23 +435,43 @@ static void restore_packet_loss_ratio(void)
...
 	restore_packet_loss_ratio();
 }
 
+ZTEST(net_socket_tcp, test_v6_send_recv_large_no_delay)
+{
+	set_packet_loss_ratio();
+	test_send_recv_large_common(1, AF_INET6);
+	restore_packet_loss_ratio();
+}
+
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:473]: Unneeded empty line.

**Review by rlubos — COMMENTED** [review]
Thanks @wirehell for contribution.

It seems that due to extra check in loopback some other test suites need MTU adjustement as well.

Regarding commit messages, the titles need to be prefixed (`drivers: net: loopback:` , `net: tcp:` `tests: net`), also please add some body to the commit messages (https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines)

**Comment by wirehell** [comment]
> It seems that due to extra check in loopback some other test suites need MTU adjustement as well.

I think the loopback MTU check exposes another issue. Tried to dig into it and for the failing test: 
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/net/socket/udp/src/main.c#L1269
No fragmentation is performed, or at least not using the MTU of the loopback.

Also stumbled upon this behavior which is most likely erroneous if both IPV6 and IPV4 is configured (it doesn't look at packet options, just config). There are no options here though, so this issue doesn't impact here.
https://github.com/zephyrproject-rtos/zephyr/blob/a2f203dd1a7a5110a3339da81e72da024be05ab5/include/zephyr/net/net_pkt.h#L691

I'll fix the things you commented on and will omit the loopback mtu check from the PR.
If I have time I might dig into the above two issues, or open issues.

**Comment by rlubos** [comment]
> No fragmentation is performed, or at least not using the MTU of the loopback.

IP fragmentation is disabled by default, so no fragmentation is expected. However, I'd rather expect the socket send call to fail if the datagram cannot fit into a single frame.

**Review by rlubos — DISMISSED** [review]
Looks reasonable.

**Comment by rlubos** [comment]
@wirehell Please check the compliance failures.

File: subsys/net/ip/tcp.c [hunk 3]
```diff
@@ -3345,13 +3345,20 @@ uint16_t net_tcp_get_supported_mss(const struct tcp *conn)
 	if (family == AF_INET) {
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:3345]: In commit "net: tcp: Fix IPv6 TCP MSS issue" could you add more information what/why is being fixed to the commit message. This way user does not necessarily go to github to see the explanation.
We can still have the pointer to the issue, so you could add
"Fixes: #56657"
to the commit message too.

**Review by rlubos — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Comment by wirehell** [comment]
> > No fragmentation is performed, or at least not using the MTU of the loopback.
> 
> IP fragmentation is disabled by default, so no fragmentation is expected. However, I'd rather expect the socket send call to fail if the datagram cannot fit into a single frame.

The test that was failing when I tried to enable MTU limit on loopback actually have fragmentation enabled: 
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/net/socket/udp/testcase.yaml#L15
I tried to trace a bit but couldn't see any fragmentation being done except on the receiving end. But I didn't look that hard as I don't have any use for that feature.

**Pull request merged**